src/KOKKOS/fix_rigid_kokkos.cpp

```diff
@@ -0,0 +1,1835 @@
+
+      auto new_xcm = domainKK->remap(prd, h, h_inv, triclinic, boxlo, boxhi,
+                                     boxlo_lambda, boxhi_lambda, prd_lambda,
+                                     xcm_ibody, imagebody_ibody, periodic_bits);
+
+      l_imagebody(ibody) = imagebody_ibody;
+      l_xcm(ibody,0) = new_xcm[0];
+      l_xcm(ibody,1) = new_xcm[1];
+      l_xcm(ibody,2) = new_xcm[2];
+    });
+  }
+
+  k_xcm.modify<DeviceType>();
+  k_imagebody.modify<DeviceType>();
+
+  image_shift_kokkos();
+
+}
+
+
+template <class DeviceType>
+void FixRigidKokkos<DeviceType>::image_shift_kokkos()
+{
+  int nlocal = atomKK->nlocal;
+
+  k_body.sync<DeviceType>();
+  k_xcmimage.sync<DeviceType>();
+
+  {
+    // Local block for Kokkos::parallel_for
+
+    imageint tdim,bdim,xdim[3];
+    auto l_image = atomKK->k_image.d_view;
+    auto l_imagebody = k_imagebody.d_view;
+    auto l_body = k_body.d_view;
+    auto l_xcmimage = k_xcmimage.d_view;
+
+    Kokkos::parallel_for(nlocal, LAMMPS_LAMBDA(const int &i) {
+        if (l_body[i] < 0) return;
+        int ibody = body[i];
+        imageint xdim[3];
+        imageint tdim = l_image[i] & IMGMASK;
+        imageint bdim = l_imagebody[ibody] & IMGMASK;
+
+        xdim[0] = IMGMAX + tdim - bdim;
+        tdim = (l_image[i] >> IMGBITS) & IMGMASK;
+        bdim = (l_imagebody[ibody] >> IMGBITS) & IMGMASK;
+        xdim[1] = IMGMAX + tdim - bdim;
+        tdim = l_image[i] >> IMG2BITS;
+        bdim = l_imagebody[ibody] >> IMG2BITS;
+        xdim[2] = IMGMAX + tdim - bdim;
+
+        l_xcmimage[i] = (xdim[2] << IMG2BITS) | (xdim[1] << IMGBITS) | xdim[0];
+    });
+  }
+  k_xcmimage.modify<DeviceType>();
+}
+
+
+
+template <class DeviceType>
+void FixRigidKokkos<DeviceType>::initial_integrate(int vflag)
+{
+  // initial integrate uses the following:
+  // vcm, fcm, fflag, angmom, torque, tflag, ex_space, ey_space,
+  // ez_space, inertia, omega, quat, x, v, f, rmass, mass, xcmimage, body
+  //
+  // initial integrate modifies the following:
+  // vcm, xcm, angmom, omega, quat, ex_space, ey_space, ez_space, x, v
+  //
+  // set_xv uses:
+  // body, xcmimage, x, v, omega, vcm, xcm, mass, f, displace
+  //
+  // set_xv modifies:
+  // x, v
+
+  k_vcm.sync<DeviceType>();
+  k_xcm.sync<DeviceType>();
+  k_fcm.sync<DeviceType>();
+
+  k_fflag.sync<DeviceType>();
+  k_tflag.sync<DeviceType>();
+  k_body.sync<DeviceType>();
+  k_xcmimage.sync<DeviceType>();
+  k_displace.sync<DeviceType>();
+
+  k_angmom.sync<DeviceType>();
+  k_omega.sync<DeviceType>();
+  k_torque.sync<DeviceType>();
+  k_inertia.sync<DeviceType>();
+
+  k_quat.sync<DeviceType>();
+
+  k_ex_space.sync<DeviceType>();
+  k_ey_space.sync<DeviceType>();
+  k_ez_space.sync<DeviceType>();
+  k_xcmimage.sync<DeviceType>();
+  k_masstotal.sync<DeviceType>();
+
+  atomKK->k_x.sync<DeviceType>();
+  atomKK->k_v.sync<DeviceType>();
+  atomKK->k_f.sync<DeviceType>();
+  atomKK->k_mass.sync<DeviceType>();
+  atomKK->k_rmass.sync<DeviceType>();
```

Comment:
These calls are redundant with line 772 (`atomKK->sync(execution_space, datamask_read);`) and should be removed.

Reply:
Done, thanks.